winch/codegen/src/codegen/context.rs

```diff
@@ -246,16 +247,23 @@ impl<'a> CodeGenContext<'a> {
         self.stack.inner_mut().truncate(truncate);
     }
 
-    /// Reset value and stack pointer to the given length
-    /// and stack pointer offset respectively.
-    pub fn reset_stack<M: MacroAssembler>(
+    /// Pops the stack pointer to ensure that it is correctly placed according
+    /// to the expectations of the destination branch.
+    ///
+    /// This function must be used when performing unconditional jumps, as the
+    /// machine stack might be left unbalanced at the jump site (e.g. due to
+    /// register spills).
+    pub fn pop_sp_for_branch<M: MacroAssembler>(
         &mut self,
+        destination: &ControlStackFrame,
         masm: &mut M,
-        stack_len: usize,
-        sp_offset: u32,
     ) {
-        masm.reset_stack_pointer(sp_offset);
-        self.drop_last(self.stack.len() - stack_len);
+        let (_, original_sp_offset) = destination.original_stack_len_and_sp_offset();
+        let current_sp_offset = masm.sp_offset();
+
```

Comment:
Can we add an assert here that `current_sp_offset >= original_sp_offset`? The invariant we want in the end is that the offsets are equal, and we handle the greater-than case above; I think we should never see the less-than case because we're going "up" the control-stack, is that right?

Reply:
Good catch. Done, thanks!

Reply:
CI disagrees, and after digging into it a bit I think that the invariant in question not always holds, legitimately:

```wat
(module
  (func (export "cont-inner") (result i32)
    (local i32)
    (local.set 0 (i32.const 0))
    (local.set 0 (i32.add (local.get 0) (loop (result i32) (loop (result i32) (br 1))))) ;; <====
    (local.set 0 (i32.add (local.get 0) (loop (result i32) (i32.ctz (br 0)))))
    (local.set 0 (i32.add (local.get 0) (loop (result i32) (i32.ctz (loop (result i32) (br 1))))))
    (local.get 0)
  )
)
```

When entering the second loop (in the second `local.set`), there are no values to spill so the current stack pointer offset is 24, same as the sp offset at the outer loop which spilled one local, then when reaching `br 1`, since the value stack only contains spilled values we need to pop that value to account for the return of the loop, which will leave the current sp at 16, effectively less than the original.

This makes sense to me, but do let me know if you see anything different.

Reply:
Ah, hmm, OK. Can we leave a comment as to how this works then? It's not obvious to me at least how this local bit of code works -- why is it OK that we can branch to a place if we have a mismatched stack -- unless the needed values are pushed somewhere else?

Reply:
Added a comment and put back an assert that goes along with the comment: in general the stack pointer can only be `word_bytes` less than stack pointer offset of destination frame, which accounts for the handling of the return values at the jump site. I hope this clarifies things!

Reply:
Sounds good!